src/network/esp/src/NetworkManager.cpp

```diff
@@ -0,0 +1,81 @@
+#include "NetworkManager.h"
+#include "NetworkConfig.h"
+
+constexpr uint8_t gs_loopRate = 100;
+
+static void task(void* context) {
+    if (context != nullptr) {
+        while (true) {
+            static_cast<NetworkManager*>(context)->execute();
+        }
+    }
+}
+
+NetworkManager::NetworkManager(ILogger& logger) :
+    m_logger(logger), m_driverTask("stm_spi_driver", tskIDLE_PRIORITY + 1, task, this) {
+    ESP_ERROR_CHECK(esp_event_handler_register(WIFI_EVENT, ESP_EVENT_ANY_ID, &eventHandler, this));
+    ESP_ERROR_CHECK(esp_event_handler_register(IP_EVENT, IP_EVENT_STA_GOT_IP, &eventHandler, this));
+    // Initialise to 0.0.0.0
+    m_ipAddress.u_addr.ip4.addr = 0;
+}
+
+void NetworkManager::initNetworkInterface() {
+    ESP_ERROR_CHECK(esp_netif_init());
+
+    wifi_init_config_t cfg = WIFI_INIT_CONFIG_DEFAULT();
+    ESP_ERROR_CHECK(esp_wifi_init(&cfg));
+
+    // Initialize default station as network interface instance. Will change in the future when
+    // enabling mesh capabilities
+    auto* networkInterface = esp_netif_create_default_wifi_sta();
+    assert(networkInterface);
+}
+
+void NetworkManager::eventHandler(void* context,
+                                  esp_event_base_t eventBase,
+                                  int32_t eventId,
+                                  void* eventData) {
+    auto* manager = static_cast<NetworkManager*>(context);
+
+    if (eventBase == WIFI_EVENT && eventId == WIFI_EVENT_STA_START) {
+        esp_wifi_connect();
+    } else if (eventBase == WIFI_EVENT && eventId == WIFI_EVENT_STA_DISCONNECTED) {
+        manager->m_logger.log(LogLevel::Error, "Network error, attempting to reconnect");
+        manager->m_state = NetworkState::DISCONNECTED;
+    } else if (eventBase == IP_EVENT && eventId == IP_EVENT_STA_GOT_IP) {
+        auto* event = (ip_event_got_ip_t*)eventData;
+        manager->m_logger.log(LogLevel::Info, "Network manager got ip:" IPSTR,
+                              IP2STR(&event->ip_info.ip));
+        manager->m_ipAddress.u_addr.ip4 = event->ip_info.ip;
+        manager->m_state = NetworkState::CONNECTED;
+    }
+}
+
+void NetworkManager::start() {
+    initNetworkInterface();
+    m_driverTask.start();
+}
+
+void NetworkManager::execute() {
+
+    switch (m_state) {
+
+    case NetworkState::CONNECTING:
+        ESP_ERROR_CHECK(esp_wifi_set_mode(NetworkConfig::getMode()));
+        ESP_ERROR_CHECK(esp_wifi_set_config(NetworkConfig::getInterface(),
+                                            NetworkConfig::getDefaultNetworkConfig()));
+        ESP_ERROR_CHECK(esp_wifi_start());
+        break;
+    case NetworkState::CONNECTED:
+        // TODO: start tcp server to monitor messages and such
+        break;
+    case NetworkState::DISCONNECTED:
+        // TODO: handle disconnection to the network, like closing server socket and such
+        break;
+        ;
+    }
```

Comment:
Maybe just add a default case and log a warning, just in case we add a state and forget to handle it